### PR TITLE
PyQT5 Backend Partial Redraw Fix

### PR DESCRIPTION
**Pull request by pshriwise**
## PR Summary

This PR is related to partial redraw buffers for QT5 on screens with `devicePixelRatio() > 1` from issue #14160. 

At first, this appeared to be a problem with the PyQT5/PySide2 `QRectangle`s being passed to `FigureCanvasQT`'s, but a deeper look showed that there was some confusion about when the `FigureCanvasQT._dpi_ratio` should be applied. This PR ensures that it is only applied when creating a partial buffer for redraw and that the original rectangle locations from PyQT5/Pyside2 are used to place redrawn region. It still addresses the off-by-one issue by using the bottom coordinate of the rectangle and shifting it by the rectangle's height.

### CODE REVIEW

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 1]
```diff
@@ -38,16 +38,17 @@ def paintEvent(self, event):
...
-        height = rect.height()
         # See documentation of QRect: bottom() and right() are off by 1, so use
         # left() + width() and top() + height().
-        bbox = Bbox(
-            [[left, self.renderer.height - (top + height * self._dpi_ratio)],
-             [left + width * self._dpi_ratio, self.renderer.height - top]])
+        rect = event.rect()
+        width = rect.width() * self._dpi_ratio
+        height = rect.height() * self._dpi_ratio
+        left, top = self.mouseEventCoords(rect.topLeft())
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5agg.py]: Can you leave a comment in the code about why we are calling this?

**Comment by tacaswell** [comment]
Thanks for digging into this!

The "simple" scaling for hi-dpi has historically been the source of a surprising a amount of grief.

**Comment by jklymak** [comment]
Do we need this change on 3.1.x?   

This prob needs someone to check that it works on different operating systems?

**Comment by tacaswell** [comment]
This should get interactive testing across several OSs and I think should be backported to both 3.1.x and 2.2.x

File: lib/matplotlib/backends/backend_qt5agg.py (hunk 1 above)
Comment (pshriwise) [reply lib/matplotlib/backends/backend_qt5agg.py]: Yeah for sure. Good call.

**Comment by pshriwise** [comment]
@tacaswell happy to do it!

@jklymak FWIW I've tested this change on Ubuntu 18.04 and OSX Mojave 10.14.3, though I'm sure more extensive testing is in order.

**Comment by pshriwise** [comment]
Is there anything else I can do here. More testing or tweaks to style?

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 2]
```diff
@@ -38,16 +38,19 @@ def paintEvent(self, event):
...
-             [left + width * self._dpi_ratio, self.renderer.height - top]])
+        rect = event.rect()
+        # scale rect dimensions using the screen dpi ratio to get
+        # correct values for the Figure coordinates (rather than QT5's coords)
+        width = rect.width() * self._dpi_ratio
+        height = rect.height() * self._dpi_ratio
+        left, top = self.mouseEventCoords(rect.topLeft())
+        # shift the "top" by the height of the image to get the
+        # correct corner for our coordinate system
+        top -= height
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_qt5agg.py]: I guess the point is that this is really bottomRight(), but with the off-by-1 correction? perhaps move the comment about the off-by-1 there, if I got this correctly.

**Comment by anntzer** [comment]
just a small clarification needed but otherwise looks good, approving conditional on it (... assuming my understanding is correct :))

**Comment by pshriwise** [comment]
It looks like CI is failing in the setup step. Is it worth a restart? Or maybe should I rebase this branch?

**Comment by anntzer** [comment]
Doc build failure is clearly spurious (this doesn't touch any docstrings).
Thanks for the PR!

**Comment by tacaswell** [comment]
This code path has diverged significantly from 2.2.x, going to defer backporting this unless there is a champion for it.

**Pull request merged**